### ISSUE TITLE
Bug in demo given in README

### ISSUE THREAD

**Issue by alanpapalia** [ethz-asl/voxgraph#32]
It appears that the TFs aren't published consistently in the ROSbag provided in the demo with the following `bash` commands 

```shell script
wget -P ~/Downloads http://robotics.ethz.ch/~asl-datasets/2020_voxgraph_arche/arche_flight1_2ms_indoor-outdoor-figure-8.zip
unzip ~/Downloads/arche_flight1_2ms_indoor-outdoor-figure-8.zip
```

When running :

```shell script
roslaunch voxgraph arche_demo.launch rosbag_path:=${HOME}/Downloads/arche_flight1_2ms_indoor-outdoor-figure-8.bag
```

I am receiving these errors:

```
[ WARN] [1594820601.500600965, 1562071535.519325233]: Lookup would require extrapolation into the future.  Requested time 1562071525.700602368 but the latest data is at time 1562071186.394310656, when looking up transform from frame [os1_lidar] to frame [odom]
[ WARN] [1594820601.552995483, 1562071535.569634331]: Waited 0.260s, but still could not get the TF from imu to odom
[ WARN] [1594820601.553026338, 1562071535.569634331]: Skipping pointcloud since the poses at time 1562071524.919065856 could not be looked up.
```

The TF tree for most of the bag appears to be as such: [frames.pdf](https://github.com/ethz-asl/voxgraph/files/4926154/frames.pdf)

**Comment by victorreijgwart** [comment]
Hi Alan, this error usually appears a few times at the start of the rosbag. After a few seconds, once everything has started up, it should go away. Does that also happen for you? Or does the error persist?
The first submap is finished after 10 seconds, so you might have to give it some time (e.g. 15s) before things start to appear in Rviz.

**Comment by alanpapalia** [comment]
The error persists throughout almost the entire bag, though it appears to find some correct transforms at the end of the file. I'm making this assumption based on the fact that I was able to get this output in RViz at the very end of the bag (after several minutes). Thank you, it is good to know that it is likely coming from the lidar_undistortion node. 

![image](https://user-images.githubusercontent.com/17442843/87609587-f7233b80-c6d0-11ea-9fb2-f127e2aa8acd.png)

**Comment by alanpapalia** [comment]
Hi, it's worth noting that when running on various parts of the KITTI dataset the errors mentioned above disappear but still most of the mesh seems to be missing. 

For KITTI data I am using the following frames, could that be the issue? It seems like suitesparse is able to formulate the problem and, for the most part, solve it. Still, at no point does there appear to be anything published as `/voxgraph_mapper/active_mesh` which seems to indicate some issue. 

```
 <arg name="pointcloud_topic" value="velodyne_points" />

input_odom_frame: "imu"
input_base_link_frame: "imu"
output_mission_frame: "mission"
output_odom_frame: "imu"
output_base_link_frame: "imu"
output_sensor_frame: "sensor_voxgraph"

```

**Comment by victorreijgwart** [comment]
What sort of computer are you running the experiments on?
The meshes should be published to `/voxgraph_mapper/combined_mesh` and `/voxgraph_mapper/separated_mesh`.
I'll boot into my work computer and post a snapshot of the tf tree here in a few minutes.

**Comment by alanpapalia** [comment]
I'm running Acer NITRO 5 AN515-53-55G9 with Intel i5 8th gen processor, Ubuntu 18.04, ROS Melodic

My apologies, I had assumed that `/voxgraph_mapper/active_mesh ` would be used but I now see that is not the case. 

This does seem unusual, as on the KITTI dataset no warnings or errors are appearing but very little mesh is being displayed, if any, even on 5 minute long bag files.

**Comment by victorreijgwart** [comment]
The computer you're running it on should definitely be fast enough.
The TF tree when running the `arche_demo.launch` on the current master branch should look like this: 
![image](https://user-images.githubusercontent.com/6238939/87726665-c241e200-c7bf-11ea-9675-a76463ebfd51.png)

**Comment by victorreijgwart** [comment]
And in Rviz it should look like this:
![output](https://user-images.githubusercontent.com/6238939/87728203-da673080-c7c2-11ea-8e2e-400dfe2a5b2e.gif)

High-quality, not sped up video here: https://u.pcloud.link/publink/show?code=XZhwPNkZyEyOer8B5UkjAytw6PbAJQceOWdk

**Comment by victorreijgwart** [comment]
The issue could be that your odom and base_link frames are set to the same value.
The base link frame should be set to the coordinate frame of the robot's body. If you're using an inertial odometry estimator, like ROVIO, it would usually estimate the pose of the robot's IMU and we would, therefore, use this.
The odom frame should be set to the fixed frame in which the odometry is expressed, for example "odom" or "world".
In other words, if your odometry estimator estimates the transformation `T_A_B`; you'd set `input_odom_frame: "A"` and `input_base_link_frame: "B"`.
I hope this helps. If anything is unclear or if the problem persists, please let me know.

**Comment by alexmillane** [comment]
If you're running the provided launchfile, all the frames _should_ be set up correctly.

I guess there could be some delay in the rosbag publisher??? I have, on occassion on slower machines, had some weird issues with playing large bags. This is a bit of a shot in the dark. Are you playing the bag off an SSD? You could try increasing the rosbag play output queue, or slowing down the play rate to give your machine more time to get things off of the disk.

**Comment by alanpapalia** [comment]
I'm running off an SSD, it seems like the bagfile is playing at a normal speed. There's an added issue I'd forgotten that is giving Compiler Error and might be indicative of the underlying issue?

```
Errors     << voxgraph:make /home/alan/catkin_ws/logs/voxgraph/build.make.012.log                   
/home/alan/catkin_ws/devel/.private/voxgraph/lib/libvoxgraph.a(loop_closure_edge_server.cpp.o): In function `voxgraph::LoopClosureEdgeServer::publishLoopClosureEdges(voxgraph::PoseGraphInterface const&, voxgraph::VoxgraphSubmapCollection const&, ros::Time const&, ros::Publisher const&, bool)':
/home/alan/catkin_ws/src/voxgraph/voxgraph/src/tools/data_servers/loop_closure_edge_server.cpp:81: undefined reference to `voxgraph::LoopClosureEdgeServer::set_unknown_covariance_entries_to_'
collect2: error: ld returned 1 exit status
make[2]: *** [/home/alan/catkin_ws/devel/.private/voxgraph/lib/voxgraph/voxgraph_mapping_node] Error 1
make[1]: *** [CMakeFiles/voxgraph_mapping_node.dir/all] Error 2
make: *** [all] Error 2
```

It seems like the issue here is some linker error where CMake isn't properly linking to `loop_closure_edge_server.h`. I've had the same issue on two separate machines, with GCC and Clang compilers. I think it might be an issue with `cs_simple`, so I'm going to try rewriting the `CMakeLists` and see if that fixes the error. 

Previously I had fixed it by redeclaring `set_unknown_covariance_entries_to` inside of `loop_closure_edge_server.cpp`. It seems like an odd issue, as the other private member variables are being accessed without error.

**Comment by alexmillane** [comment]
Any progress here Alan?

**Comment by alanpapalia** [comment]
Hey Alex, thanks for checking in. We've had success installing and running on a different computer (the 3rd computer tried). Sometime soon I think I'm going to try to check all the different software versions and see if that's the issue. It still seems like there's a difference in how it's getting compiled, as there are some odd linker errors on the first two computers attempted, but it's not obvious why that's the case. 

I'll let you know if I find anything!

**Comment by alexmillane** [comment]
Hmm weird. Let us know if you find anything. Probably time for us to set up some CI honestly.

**Comment by LXYYY** [comment]
hi, I am running voxgraph in both ubuntu 16 and 18 dockers, and I have the same issue with Alan, firstly, the set_unknown_covariance_entries_to_ undefined issue, and secondly the optimization going really slow. But the most weirdly, yesterday when I ran it on a ubuntu18 docker, everything works perfectly, optimization very fast, and no that undefined issue. Yesterday It's like this:
![image](https://user-images.githubusercontent.com/15899704/90137588-6a8b9c00-dda8-11ea-835d-736264a349ec.png)

and today, both in 16 and 18 docker, it looks like this:
![image](https://user-images.githubusercontent.com/15899704/90137664-88590100-dda8-11ea-9826-6142e582b1ad.png)

the only thing different I can think up is yesterday docker base image is a cuda image with ros installed on top, or a origin ros image, but today I am using a tensorflow-gpu base image. 

Do you have any idea? I will try roll back the environment to see the performance

**Comment by alanpapalia** [comment]
As you were implying, based on some tests that I ran, it definitely seemed to be an issue where running Voxgraph was using all of the system resources (at least that appeared to be the case for the two machines I had issues with) and as a result couldn't read the rosbag quickly enough. Because the computation seemed too heavy for the platforms I eventually wanted to work with, I have since moved to a different dense SLAM system.

I hadn't explored all of the different configurations and potential impacts they may have on performance (besides simple compiler optimizations). Let me know how the different docker images work for you!

Can I ask what hardware you're running this on, for comparison?

**Comment by LXYYY** [comment]
It's a desktop in my lab, with a AMD cpu I can't remember the model since I'm not very familiar with amd cpu, and I'm not in lab now, and a 1660 Super gpu. Anyway it's a lab desktop, and definitely powerful enough.

**Comment by alexmillane** [comment]
The paper results were generated on an i7 NUC, so a modern laptop should be sufficient.

With that said, it does look like the data isn't getting integrated fast enough. Could you check that you're building in release mode?
```
cd catkin_ws
catkin config
```
you should see:
`Additional CMake Args:       -DCMAKE_BUILD_TYPE=Release`

**Comment by LXYYY** [comment]
Hi Alex, I have never built anything in debug during setting up these days. I don't think it's a build type issue. Now I suspect the tensor flow source code building process might changed some system lib version, etc. and the image I used today is a devel one, which may have more changes on the environment than runtime ones. I'll try more tomorrow

**Comment by LXYYY** [comment]
Let's say maybe protobuf? Voxblox is using block and layer as proto, and maybe it's been changed somehow during the some process like tensorflow building? I know little about protobuf, this guess may be just nonsense.

**Comment by alexmillane** [comment]
We're only using protobuf to serialize and deserialize maps for writing to disk in response to a save request. So it shouldn't be that (although protobuf is also usually my first suspicion ;) ).

**Comment by LXYYY** [comment]
I rolled back my dockerfile, and in some commit the problem was gone, and every works perfectly. Then I manually added steps back to dockerfile one by one trying to reproduce the problem, but it has not shown up ever. I still have no idea about the reason yet. Alan, if you still want to give voxgraph a try, you can try my dockerfile, [voxgraph_docker](git@github.com:LXYYY/voxgraph_docker.git), use master branch, it has ros-kinetic-perception, ubuntu16, tf1.6-gpu, etc. and remember to run it with nvidia docker environment.

**Comment by victorreijgwart** [comment]
Thank you for sharing your docker files @LXYYY, and glad to hear it's working for you now.
Hopefully, we'll find time to replicate and narrow down the cause of this problem in the future. If others run into the same issue in the meantime, I'd refer them to your repository for docker setups if that's ok with you.
